File: services/celery-gateway/src/celery_gateway/routers/tasks.py
from __future__ import annotations

import asyncio
import json
import logging
from typing import Any

from celery.result import AsyncResult
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.responses import JSONResponse

from ..celery_app import app as celery_app
from ..middleware.auth import require_auth
from ..models.tasks import (
    FrontendActiveTask,
    FrontendRegisteredTasksResult,
    FrontendSendTaskRequest,
    FrontendSendTaskResponse,
    FrontendTaskHistoryItem,
    FrontendTaskPayload,
    FrontendTaskStatusResponse,
    RevokeRequest,
    RevokeResponse,
)
from ..services.celery_redis import (
    get_celery_task_status,
    get_task_payloads,
    send_celery_task,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/tasks", tags=["tasks"])


# ------------------------------------------------------------------
# Frontend-facing endpoints (previously in Hono server)
# ------------------------------------------------------------------


@router.get("/active", response_model=list[FrontendActiveTask])
async def frontend_active_tasks(request: Request) -> Any:
    cache = request.app.state.celery_cache
    return await cache.get("active-tasks")


@router.get("/history", response_model=list[FrontendTaskHistoryItem])
async def frontend_task_history(
    request: Request,
    limit: int = Query(50, ge=1, le=200),
) -> Any:
    cache = request.app.state.celery_cache
    tasks: list[dict[str, Any]] = await cache.get("task-history")
    return tasks[:limit]


@router.get("/registered", response_model=FrontendRegisteredTasksResult)
async def frontend_registered_tasks(request: Request) -> Any:
    cache = request.app.state.celery_cache
    return await cache.get("registered-tasks")


@router.get("/payloads", response_model=list[FrontendTaskPayload])
async def frontend_task_payloads(
    name: str = Query(..., description="Task name"),
) -> list[dict[str, Any]]:
    return await get_task_payloads(name)


@router.post(
    "/send",
    response_model=FrontendSendTaskResponse,
    dependencies=[Depends(require_auth)],
)
async def frontend_send_task(body: FrontendSendTaskRequest) -> JSONResponse:
    args: list[Any] = json.loads(body.args)
    kwargs: dict[str, Any] = json.loads(body.kwargs)
    queue = body.queue or "celery"

    # Try Celery app first, fall back to direct Redis
    try:
        loop = asyncio.get_running_loop()
        result = await loop.run_in_executor(
            None,
            lambda: celery_app.send_task(
                body.task_name,
                args=args,
                kwargs=kwargs,
                queue=queue,
                countdown=body.countdown,
                eta=body.eta,
                priority=body.priority,
            ),
        )
        return JSONResponse({"taskId": result.id})
    except Exception:
        try:
            task_id = await send_celery_task(body.task_name, args, kwargs, queue)
            return JSONResponse({"taskId": task_id})
        except Exception as exc:
            return JSONResponse(
                {"error": f"Failed to send task: {exc}"}, status_code=500
            )


@router.post(
    "/{task_id}/revoke",
    response_model=RevokeResponse,
    dependencies=[Depends(require_auth)],
)
async def frontend_revoke_task(
    task_id: str,
    body: RevokeRequest | None = None,
) -> JSONResponse:
    revoke_opts = body or RevokeRequest()
    try:
        loop = asyncio.get_running_loop()
        await loop.run_in_executor(
            None,
            lambda: celery_app.control.revoke(
                task_id,
                terminate=revoke_opts.terminate,
                signal=revoke_opts.signal,
            ),
        )
        return JSONResponse({"task_id": task_id, "revoked": True})
    except Exception:
        return JSONResponse({"error": "Failed to revoke task"}, status_code=503)


@router.get("/{task_id}/status", response_model=FrontendTaskStatusResponse)
async def frontend_task_status(task_id: str) -> Any:
    # Try Celery backend first, fall back to Redis
    try:
        loop = asyncio.get_running_loop()
        result: AsyncResult = await loop.run_in_executor(
            None, lambda: AsyncResult(task_id, app=celery_app)
        )

        return {
            "status": result.status,
            "result": result.result if result.status != "PENDING" else None,
        }
    except Exception:
        try:
            redis_result = await get_celery_task_status(task_id)
            if not redis_result:
                raise HTTPException(status_code=404, detail="Task not found")
            return {
                "status": redis_result["status"],
                "result": redis_result.get("result"),
            }
        except HTTPException:
            raise
        except Exception:
            raise HTTPException(status_code=404, detail="Task not found")

File: packages/server/src/routes/tasks.ts
import { Hono } from "hono";
import { getCachedActiveTasks, getCachedTaskHistory, getCachedRegisteredTasks } from "../lib/cache.js";
import { getTaskPayloads } from "../lib/celery.js";
import { gatewayRevokeTask, gatewaySendTask, gatewayGetTaskStatus } from "../lib/celery-gateway.js";
import { sendCeleryTask, getCeleryTaskStatus } from "../lib/celery.js";

const app = new Hono();

// GET /api/tasks/active
app.get("/active", async (c) => {
  const tasks = await getCachedActiveTasks();
  return c.json(tasks);
});

// GET /api/tasks/history
app.get("/history", async (c) => {
  const limit = Math.min(
    Math.max(Number(c.req.query("limit")) || 50, 1),
    200,
  );
  const tasks = await getCachedTaskHistory();
  return c.json(tasks.slice(0, limit));
});

// GET /api/tasks/registered
app.get("/registered", async (c) => {
  const data = await getCachedRegisteredTasks();
  return c.json(data);
});

// GET /api/tasks/payloads
app.get("/payloads", async (c) => {
  const name = c.req.query("name");
  if (!name) {
    return c.json({ error: "name is required" }, 400);
  }

  try {
    const payloads = await getTaskPayloads(name);
    return c.json(payloads);
  } catch (err) {
    return c.json(
      { error: `Failed to fetch payloads: ${err}` },
      500
    );
  }
});

// POST /api/tasks/:id/revoke
app.post("/:id/revoke", async (c) => {
  const id = c.req.param("id");

  try {
    const body = await c.req.json().catch(() => ({}));
    const terminate = body?.terminate ?? false;
    const ALLOWED_SIGNALS = ["SIGTERM", "SIGKILL"];
    const signal = ALLOWED_SIGNALS.includes(body?.signal) ? body.signal : "SIGTERM";
    const result = await gatewayRevokeTask(id, terminate, signal);
    return c.json(result);
  } catch {
    return c.json(
      { error: "Gateway unavailable" },
      503
    );
  }
});

// POST /api/tasks/send
app.post("/send", async (c) => {
  const body = await c.req.json().catch(() => null);
  if (!body) {
    return c.json({ error: "Invalid JSON body" }, 400);
  }

  const taskName = body.taskName as string;
  const queue = (body.queue as string) || "celery";
  const argsRaw = (body.args as string) || "[]";
  const kwargsRaw = (body.kwargs as string) || "{}";
  const countdown = body.countdown != null ? parseFloat(body.countdown) : null;
  const eta = body.eta || null;
  const priority = body.priority != null ? parseInt(body.priority, 10) : null;

  if (!taskName) {
    return c.json({ error: "Task name is required" }, 400);
  }

  if (!/^[\w.]+$/.test(taskName)) {
    return c.json({ error: "Invalid task name format" }, 400);
  }

  let args: unknown[];
  try {
    args = JSON.parse(argsRaw);
    if (!Array.isArray(args)) {
      return c.json({ error: "Args must be a JSON array" }, 400);
    }
  } catch {
    return c.json({ error: "Invalid JSON for args" }, 400);
  }

  let kwargs: Record<string, unknown>;
  try {
    kwargs = JSON.parse(kwargsRaw);
    if (typeof kwargs !== "object" || Array.isArray(kwargs) || kwargs === null) {
      return c.json({ error: "Kwargs must be a JSON object" }, 400);
    }
  } catch {
    return c.json({ error: "Invalid JSON for kwargs" }, 400);
  }

  // Try gateway first, fall back to direct Redis
  try {
    const res = await gatewaySendTask({
      task_name: taskName,
      args,
      kwargs,
      queue,
      countdown,
      eta,
      priority,
    });
    return c.json({ taskId: res.task_id });
  } catch {
    // Gateway unavailable — fall back to direct Redis
    try {
      const taskId = await sendCeleryTask(taskName, args, kwargs, queue);
      return c.json({ taskId });
    } catch (err) {
      return c.json({ error: `Failed to send task: ${err}` }, 500);
    }
  }
});

// GET /api/tasks/:id/status
app.get("/:id/status", async (c) => {
  const taskId = c.req.param("id");

  // Try gateway first, fall back to direct Redis
  try {
    const res = await gatewayGetTaskStatus(taskId);
    return c.json({ status: res.status, result: res.result });
  } catch {
    try {
      const result = await getCeleryTaskStatus(taskId);
      if (!result) return c.json(null);
      return c.json({ status: result.status, result: result.result });
    } catch {
      return c.json(null);
    }
  }
});

export default app;
